Add support for K66F and Segger IP switch board

parthitce opened the pull request:
Add support for K66F and [Segger IP switch board](https://www.segger.com/evaluate-our-software/segger/embosip-switch-board/) in existing K64F SoC tree.
Switch support for Zephyr is in progress: https://github.com/zephyrproject-rtos/zephyr/issues/22061

HAL Depends: https://github.com/zephyrproject-rtos/hal_nxp/pull/31

Commit:
soc: arm: nxp: kinetis: k6x: add K66F support

Add MK66F18 series support with MK66FN2M0VMD18 part.

Signed-off-by: Parthiban Nallathambi <pn@denx.de>

Commit:
boards: arm: Add ip_k66f board support

IP ethernet switch board includes K66F MCU from NXP
and Micrel/Microchip KSZ8794CNX switch. Board support
includes basic GPIO LED and common functions.

Switch function needs support either based on DSA or
relative feature in Zephyr which is tracked with issue
22061.

Signed-off-by: Parthiban Nallathambi <pn@denx.de>

jukkar commented on boards/arm/ip_k66f/doc/index.rst:
```diff
@@ -0,0 +1,175 @@
+A debug probe is used for both flashing and debugging the board. This board is
+configured by default to use the :ref:`opensda-jlink-onboard-debug-probe`.
+
+Option 1: :ref:`opensda-jlink-onboard-debug-probe`
```
Do we need to mention `Option 1` here, as there is no other options mentioned in this file.

parthitce replied:
Thanks, fixed.

jukkar commented on boards/arm/ip_k66f/doc/index.rst:
```diff
@@ -0,0 +1,175 @@
+does provide an OpenSDA J-Link Board-Specific Firmware for this board, however
+it is not compatible with the DAPLink bootloader.
+
+Add the argument ``-DOPENSDA_FW=jlink`` when you invoke ``west build`` to
```
Perhaps this option could be added as a default for the board if it is always the same?

BTW, is this the normal way to set the runner, I have used the `west flash -r jlink ....` for this. @mbolivar what is the suggested way the runner should be set?

parthitce replied:
@jukkar Am not sure how to make this default. Adding `set(ZEPHYR_BOARD_FLASH_RUNNER jlink)` to board's Cmake doesn't have any impact.

mbolivar replied:
Here is an example from frdm_k64f showing how to use `board_set_flasher_ifnset()` and `board_set_debugger_ifnset()` to set the default runners while still respecting any choice the user makes at CMake time.

```
set_ifndef(OPENSDA_FW daplink)

if(OPENSDA_FW STREQUAL jlink)
  board_set_debugger_ifnset(jlink)
  board_set_flasher_ifnset(jlink)
elseif(OPENSDA_FW STREQUAL daplink)
  board_set_debugger_ifnset(pyocd)
  board_set_flasher_ifnset(pyocd)
endif()

board_runner_args(jlink "--device=MK64FN1M0xxx12")
board_runner_args(pyocd "--target=k64f")

include(${ZEPHYR_BASE}/boards/common/pyocd.board.cmake)
include(${ZEPHYR_BASE}/boards/common/jlink.board.cmake)
include(${ZEPHYR_BASE}/boards/common/openocd.board.cmake)
```

Note that if you don't use a `board_set_foo_ifnset()` function, then the first `include` line will set the defaults, so for example pyocd.board.cmake would set the default runner to pyocd if not already set.

I personally don't like the continued use of the `OPENSDA_FW` variable as it's vendor specific and there is a more general mechanism available via the runner configuration in CMake, but that is @MaureenHelm 's decision since she maintains these SoCs.

parthitce replied:
@mbolivar Thanks you. Updated the document and board.cmake.

jukkar commented on boards/arm/ip_k66f/pinmux.c:
```diff
@@ -0,0 +1,43 @@
+	struct device *porte =
+		device_get_binding(CONFIG_PINMUX_MCUX_PORTE_NAME);
+#endif
+
```
We are defining various port[a-e] variables above, but only porta is used below. Is something missing here?

parthitce replied:
@jukkar Yes, for now only port A is used. B, E will be used when the switch support is added. C, D can be removed as there isn't any connections. I will remove it. Thanks.

galak replied:
As @jukkar asked, lets remove PORTB & E for now.  Add them back when you have the additional code for the switch.

parthitce commented:
@jukkar Could you please merge the [HAL PR](https://github.com/zephyrproject-rtos/hal_nxp/pull/31) to check for the bot compilation? Thanks.

chrta commented:
> @jukkar Could you please merge the [HAL PR](https://github.com/zephyrproject-rtos/hal_nxp/pull/31) to check for the bot compilation? Thanks.

You can fix compilation by letting west.yml point to your open pr, see https://docs.zephyrproject.org/latest/guides/modules.html#changes-to-existing-modules

```diff
diff --git a/west.yml b/west.yml
index beb582777b..075e927680 100644
--- a/west.yml
+++ b/west.yml
@@ -86,7 +86,7 @@ manifest:
       revision: 4bff01084d225996e4aae84b98be5969e2f9f33d
       path: tools/net-tools
     - name: hal_nxp
-      revision: 3e3f20e7de7606d735cdeefe21d5f2a51bf06095
+      revision: pull/31/head
       path: modules/hal/nxp
     - name: open-amp
       revision: 9b591b289e1f37339bd038b5a1f0e6c8ad39c63a
```

MaureenHelm commented on soc/arm/nxp_kinetis/k6x/Kconfig.defconfig.mk64f12:
```diff
@@ -8,6 +8,34 @@ if SOC_MK64F12
 config SOC
 	default "mk64f12"
 
+config K6X_CORE_CLOCK_DIVIDER
```
I'm fine with the renaming, but please move these configs back to `soc/arm/nxp_kinetis/k6x/Kconfig.soc`. New symbols shouldn't be defined in defconfig files

parthitce replied:
@MaureenHelm Thanks, rearranged it now. As the defaults are maintained for K64, for `CONFIG_K6X_FLEXBUS_CLOCK_DIVIDER=2` this values is moved to board defconfig.

MaureenHelm commented on soc/arm/nxp_kinetis/k6x/soc.c:
```diff
@@ -120,7 +120,7 @@ static ALWAYS_INLINE void clock_init(void)
  */
 
-static int fsl_frdm_k64f_init(struct device *arg)
+static int k6x_init(struct device *arg)
```
This change isn't related to clock defines. Please move it to commit `soc: arm: nxp: kinetis: k6x: add K66F support`

parthitce replied:
@MaureenHelm Thanks, resolved.

MaureenHelm commented on soc/arm/nxp_kinetis/k6x/Kconfig.defconfig.mk66f18:
```diff
@@ -0,0 +1,112 @@
+if SOC_MK66F18
+
+config SOC
+	default "mk66f18"
```
Please indent with a tab instead of two spaces

parthitce replied:
@MaureenHelm It was in tab, but just re-confirmed it. Thanks.

MaureenHelm commented on soc/arm/nxp_kinetis/k6x/Kconfig.defconfig.mk66f18:
```diff
@@ -0,0 +1,112 @@
+	default 5
+	help
+	  This option specifies the divide value for the K6X flash clock from the
+	  system clock.
```
Don't introduce new config symbols in a defconfig file

parthitce replied:
@MaureenHelm Thanks, resolved.

MaureenHelm commented on soc/arm/nxp_kinetis/k6x/Kconfig.defconfig.mk66f18:
```diff
@@ -0,0 +1,112 @@
+endif # GPIO
+
+config SPI
+	default n
```
Boolean configs are implicitly default n, so you can remove these two lines

parthitce replied:
@MaureenHelm Thanks, removed it.

MaureenHelm commented on boards/arm/ip_k66f/Kconfig.board:
```diff
@@ -0,0 +1,7 @@
+config BOARD_IP_K66F
+	bool "Segger IP-K66F"
+	depends on SOC_SERIES_KINETIS_K6X
+	select SOC_PART_NUMBER_MK66FN2M0VMD18
```
Please intent by one tab instead of two spaces

parthitce replied:
@MaureenHelm It was in tab, but just re-confirmed it. Thanks.

MaureenHelm commented on boards/arm/ip_k66f/Kconfig.defconfig:
```diff
@@ -0,0 +1,43 @@
+if BOARD_IP_K66F
+
+config BOARD
+	default "ip_k66f"
```
Please intent by one tab instead of two spaces

parthitce replied:
@MaureenHelm It was in tab, but just re-confirmed it. Thanks.

MaureenHelm commented on boards/arm/ip_k66f/doc/index.rst:
```diff
@@ -0,0 +1,168 @@
+Serial Port
+===========
+
+The K66F SoC has six UARTs. None is used.
```
How do you get a console? Segger RTT?

parthitce replied:
@MaureenHelm Yes, console is only possible with Segger RTT with below settings as there is no hardware UART exposed for console or other purposes.

Device name. Default: MK66FN2M0XXX18 > 
Target interface. > SWD
Interface speed [kHz]. Default: 4000 kHz > 
RTT Control Block address. Default: auto-detection > 
RTT Channel name or index. Default: channel 1 > 0

MaureenHelm replied:
then add `CONFIG_USE_SEGGER_RTT=y` to `boards/arm/ip_k66f/ip_k66f_defconfig`

parthitce replied:
@MaureenHelm Done, thanks

MaureenHelm reviewed (changes requested):
Thanks @parthitce for the contribution!

Commit:
soc: arm: nxp: kinetis: k6x: move clock defines

Rename the clock defines from k64 to K6X. This allows
inclusion of K66 series MCU's support without much
re-defines. Clock divider defaults to K64 series.

Signed-off-by: Parthiban Nallathambi <pn@denx.de>

parthitce commented:
> > @jukkar Could you please merge the [HAL PR](https://github.com/zephyrproject-rtos/hal_nxp/pull/31) to check for the bot compilation? Thanks.
> 
> You can fix compilation by letting west.yml point to your open pr, see https://docs.zephyrproject.org/latest/guides/modules.html#changes-to-existing-modules
> 
> ```diff
> diff --git a/west.yml b/west.yml
> index beb582777b..075e927680 100644
> --- a/west.yml
> +++ b/west.yml
> @@ -86,7 +86,7 @@ manifest:
>        revision: 4bff01084d225996e4aae84b98be5969e2f9f33d
>        path: tools/net-tools
>      - name: hal_nxp
> -      revision: 3e3f20e7de7606d735cdeefe21d5f2a51bf06095
> +      revision: pull/31/head
>        path: modules/hal/nxp
>      - name: open-amp
>        revision: 9b591b289e1f37339bd038b5a1f0e6c8ad39c63a
> ```

@chrta Thanks, I wasn't aware of this option. Just pushed it for checking.

parthitce commented:
Re-based to master. Can this be merged for next release  @galak @MaureenHelm @nashif?

galak commented on boards/arm/ip_k66f/doc/index.rst:
```diff
@@ -0,0 +1,171 @@
+Serial Port
+===========
+
+The K66F SoC has six UARTs. None is used.
```
nit: 'None of them are used'.

parthitce replied:
Fixed, thanks

galak commented on boards/arm/ip_k66f/ip_k66f.dts:
```diff
@@ -0,0 +1,90 @@
+};
+
+&flash0 {
+	reg = <0 0x200000>;
```
Why is `reg` modified here?

parthitce replied:
To adapt the flash size. I will create K66 dtsi for the flash size.

parthitce replied:
Moved it to k66 dtsi. Thanks.

galak commented on boards/arm/ip_k66f/support/openocd.cfg:
```diff
@@ -0,0 +1,5 @@
+
```
We should probably add a copyright/license on this file.  I know most of the others don't have it.

parthitce replied:
@galak Added the SPDX-License-Identifier. Thanks.

galak reviewed (dismissed):
Some minor cleanups required.  I'm guessing we probably need to introduce a K66 soc dtsi file.  (for at least having the flash size

MaureenHelm commented on boards/arm/ip_k66f/ip_k66f.dts:
```diff
@@ -0,0 +1,90 @@
+	leds {
+		compatible = "gpio-leds";
+		red0_led: led_0 {
+			gpios = <&gpioa 8 GPIO_INT_ACTIVE_HIGH>;
```
Replace `GPIO_INT_ACTIVE_HIGH` with `GPIO_ACTIVE_HIGH`. Do this for led_2 as well

parthitce replied:
Fixed, thanks.

MaureenHelm commented on boards/arm/ip_k66f/ip_k66f.yaml:
```diff
@@ -0,0 +1,17 @@
+  - counter
+  - gpio
+  - nvs
+  - watchdog
```
add `usb_device`

parthitce replied:
Fixed, thanks.

MaureenHelm reviewed (changes requested):
Thanks for making the requested changes. I found a few more things, but it's getting close.

parthitce reviewed (commented):
> Some minor cleanups required. I'm guessing we probably need to introduce a K66 soc dtsi file. (for at least having the flash size

@galak Introduced k66.dtsi and included it with 2MB flash. Thanks.

parthitce commented:
@galak Although ethernet is not enabled in dts and yamil, sanity check still performs testing related to ethernet, networking and failed. Is there a way to overcome this case? Thanks.

MaureenHelm commented on west.yml:
```diff
@@ -86,7 +86,7 @@ manifest:
       path: tools/net-tools
     - name: hal_nxp
-      revision: 3e3f20e7de7606d735cdeefe21d5f2a51bf06095
+      revision: pull/31/head
```
Please update to the new SHA now that zephyrproject-rtos/hal_nxp#31 has been merged

MaureenHelm reviewed (changes requested):
@galak please revisit

MaureenHelm commented:
> @galak Although ethernet is not enabled in dts and yamil, sanity check still performs testing related to ethernet, networking and failed. Is there a way to overcome this case? Thanks.

The problem is that you enable the ethernet driver in Kconfig but not in dts. If you don't want to enable ethernet, I recommend this change:
```diff
diff --git a/soc/arm/nxp_kinetis/k6x/Kconfig.defconfig.mk66f18 b/soc/arm/nxp_kinetis/k6x/Kconfig.defconfig.mk66f18
index 26a543094b..5c12a7b8ea 100644
--- a/soc/arm/nxp_kinetis/k6x/Kconfig.defconfig.mk66f18
+++ b/soc/arm/nxp_kinetis/k6x/Kconfig.defconfig.mk66f18
@@ -42,13 +42,6 @@ config SPI_MCUX_DSPI

 endif # SPI

-if NET_L2_ETHERNET
-
-config ETH_MCUX
-       default y
-
-endif # NET_L2_ETHERNET
-
 if ENTROPY_GENERATOR

 config ENTROPY_MCUX_RNGA
```
